// diff2html-cli/build/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
function parseArgv(argv) {
    var diff2htmlOptions = {
        inputFormat: "diff",
        outputFormat: argv.style === "side" ? "side-by-side" : "line-by-line",
        showFiles: argv.summary !== "hidden",
        diffStyle: argv.diffStyle,
        matching: argv.matching,
        matchWordsThreshold: argv.matchWordsThreshold,
        matchingMaxComparisons: argv.matchingMaxComparisons
    };
    var defaultWrapperTemplate = path.resolve(__dirname, "..", "template.html");
    var configuration = {
        showFilesOpen: argv.summary === "open" || false,
        synchronisedScroll: argv.synchronisedScroll,
        highlightCode: argv.highlightCode,
        formatType: argv.format,
        outputDestinationType: argv.output,
        outputDestinationFile: argv.file,
        inputSource: argv.input,
        diffyType: argv.diffy,
        htmlWrapperTemplate: argv.htmlWrapperTemplate || defaultWrapperTemplate,
        ignore: argv.ignore || []
    };
    return [diff2htmlOptions, configuration];
}
exports.parseArgv = parseArgv;
