// diff2html-cli/build/http-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var request = require("request");
function put(url, payload) {
    return new Promise(function (resolve, reject) {
        request({
            url: url,
            method: "PUT",
            headers: {},
            body: payload,
            json: true
        })
            .on("response", function (response) {
            response.on("data", function (body) {
                try {
                    var jsonObj = JSON.parse(body.toString("utf8"));
                    if (jsonObj) {
                        return resolve(jsonObj);
                    }
                    else if (jsonObj.error !== undefined && typeof jsonObj.error === "string") {
                        return reject(new Error(jsonObj.error));
                    }
                    else {
                        return reject(new Error("Failed to read response.\n                Body:\n                " + body.toString("utf8")));
                    }
                }
                catch (err) {
                    return reject(err);
                }
            });
        })
            .on("error", function (err) { return reject(err); });
    });
}
exports.put = put;
