// diff2html-cli/build/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var childProcess = require("child_process");
var fs = require("fs");
function exists(filePath) {
    try {
        return fs.existsSync(filePath);
    }
    catch (ignore) {
        return false;
    }
}
exports.exists = exists;
function readFile(filePath) {
    return fs.readFileSync(filePath, "utf8");
}
exports.readFile = readFile;
function readStdin() {
    return new Promise(function (resolve) {
        var content = "";
        process.stdin.resume();
        process.stdin.on("data", function (buf) {
            content += buf.toString("utf8");
        });
        process.stdin.on("end", function () { return resolve(content); });
    });
}
exports.readStdin = readStdin;
function writeFile(filePath, content) {
    return fs.writeFileSync(filePath, content);
}
exports.writeFile = writeFile;
function execute(cmd) {
    return childProcess.execSync(cmd).toString("utf8");
}
exports.execute = execute;
function replaceExactly(value, searchValue, replaceValue) {
    return value.replace(searchValue, function () { return replaceValue; });
}
exports.replaceExactly = replaceExactly;
